consensus/state.go

```diff
@@ -1819,15 +1824,24 @@ func (cs *State) defaultSetProposal(proposal *types.Proposal) error {
 	if cs.ProposalBlockParts == nil {
 		cs.ProposalBlockParts = types.NewPartSetFromHeader(proposal.BlockID.PartSetHeader)
 	}
+	if cs.ProposalBlockRows == nil {
+		cs.ProposalBlockRows = types.NewRowSetFromHeader(proposal.DAHeader)
+	}
+	if cs.ProposalBlockRows.TotalSize() > int(cs.state.ConsensusParams.Block.MaxBytes) {
+		return fmt.Errorf("propasal for block exceeding maximum block size (%d > %d)",
+			cs.ProposalBlockRows.TotalSize(), cs.state.ConsensusParams.Block.MaxBytes,
+		)
+	}
```

Comment:
Now, instead of receiving parts, we can know in advance if the block exceeds the maximum size just by looking at `Proposal.DAHeader`, which is good.

Reply:
I thought we were planning on eventually removing the DA header from proposal since it's not strictly necessary. If we do remove it then we need another way of determining block size, which fortunately is deteminable from the [`Header.availableDataOriginalSharesUsed`](https://github.com/celestiaorg/celestia-specs/blob/ec98170398dfc6394423ee79b00b71038879e211/src/specs/data_structures.md#header) field.

> The size of the original data square, availableDataOriginalSquareSize, isn't explicitly declared in the block header. Instead, it is implicitly computed as the smallest power of 2 whose square is at least availableDataOriginalSharesUsed (in other words, the smallest power of 4 that is at least availableDataOriginalSharesUsed).

Reply:
@adlerjohn, in current tm's implementation(using PartSetHeader) a node checks for blocks size every time it receives a chunk(part). In the above implementation(using DAHeader) we only need a Proposal and zero chunks(rows) to understand that block exceeds the limit. In case we want to rely on `Header.availableDataOriginalSharesUsed`, we would need to receive the entire block to compute the field and only after rejecting it.